File: src/api/discussion-messages.ts
import { RequestFN } from "~/core/request-function";
import { decodeDiscussionMessages } from "~/decoders/discussion-messages";
import { type Discussion, DiscussionMessages, type SessionHandle, TabLocation } from "~/models";

/**
 * Fetches the messages and writes them in the discussion.
 * By default it won't mark the messages as read even after fetching them.
 *
 * You can change this behavior by setting `markAsRead` to `true`.
 * There's no other way to mark the messages as read.
 *
 * @param markAsRead Whether to mark the messages as read after fetching them.
 */
export const discussionMessages = async (session: SessionHandle, discussion: Discussion, markAsRead = false): Promise<DiscussionMessages> => {
  const request = new RequestFN(session, "ListeMessages", {
    _Signature_: { onglet: TabLocation.Discussions },

    donnees: {
      listePossessionsMessages: discussion.possessions,
      marquerCommeLu: markAsRead,
      nbMessagesVus: 0 // fetch all messages
    }
  });

  const response = await request.send();
  const messages = decodeDiscussionMessages(response.data.donnees, session);

  if (!discussion.messages) // setup the reference
    discussion.messages = messages;
  else // mutate the reference
    Object.assign(discussion.messages, messages);

  return messages;
};

File: library/src/commonMain/kotlin/ink/literate/pawnote/api/discussion-messages.kt
package ink.literate.pawnote.api

import ink.literate.pawnote.core.RequestFN
import ink.literate.pawnote.decoders.decodeDiscussionMessages
import ink.literate.pawnote.models.Discussion
import ink.literate.pawnote.models.DiscussionMessages
import ink.literate.pawnote.models.SessionHandle
import ink.literate.pawnote.models.TabLocation
import kotlinx.serialization.encodeToString
import kotlinx.serialization.json.*

/**
 * Fetches the messages and writes them in the discussion.
 * By default it won't mark the messages as read even after fetching them.
 *
 * You can change this behavior by setting `markAsRead` to `true`.
 * There's no other way to mark the messages as read.
 *
 * @param markAsRead Whether to mark the messages as read after fetching them.
 */
suspend fun discussionMessages (session: SessionHandle, discussion: Discussion, markAsRead: Boolean = false): DiscussionMessages {
    val request = RequestFN(session.information, "ListeMessages", Json.encodeToString(
        buildJsonObject {
            putJsonObject("_Signature_") {
                put("onglet", TabLocation.Discussions.code)
            }

            putJsonObject("donnees") {
                put("listePossessionsMessages", discussion.possessions)
                put("marquerCommeLu", markAsRead)
                put("nbMessagesVus", 0)
            }
        }
    ))

    val response = request.send()
    val messages = decodeDiscussionMessages(Json.parseToJsonElement(response.data).jsonObject["donnees"]!!.jsonObject, session)

    discussion.messages = messages

    return messages
}